synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+			ID:    addCounter(),
+			Event: requestPluginHealth,
+		},
+	}
+
+	resp := new(scheme.ResponsePluginHealth)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// Scan returns the list of devices that Synse knows about and can read
+// from/write to via the configured plugins.
+// It can be filtered to show only those devices which match a set
+// of provided tags by using ScanOptions.
+func (c *websocketClient) Scan(opts scheme.ScanOptions) (*[]scheme.Scan, error) {
+	req := scheme.RequestScan{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestScan,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseDeviceSummary)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// Tags returns the list of all tags currently associated with devices.
+// If no TagsOptions is specified, the default tag namespace will be used.
+func (c *websocketClient) Tags(opts scheme.TagsOptions) (*[]string, error) {
+	req := scheme.RequestTags{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestTags,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseTags)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data.Tags, nil
+}
+
+// Info returns the full set of meta info and capabilities for a specific
+// device.
+func (c *websocketClient) Info(id string) (*scheme.Info, error) {
+	req := scheme.RequestInfo{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestInfo,
+		},
+		Data: scheme.WriteData{
+			Device: id,
+		},
+	}
+
+	resp := new(scheme.ResponseDevice)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// Read returns data from devices which match the set of provided tags
+// using ReadOptions.
+func (c *websocketClient) Read(opts scheme.ReadOptions) (*[]scheme.Read, error) {
+	req := scheme.RequestRead{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestRead,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseReading)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// ReadDevice returns data from a specific device.
+// It is the same as Read() where the label matches the device id tag
+// specified in ReadOptions.
+// NOTE - this method is not applicable for websocket client.
```

Comment:
this should be supported. we could ultimately just forward it to the `Read` method, but I believe we should support this